trie/utils/verkle.go

```diff
@@ -0,0 +1,350 @@
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package utils
+
+import (
+	"encoding/binary"
+	"sync"
+
+	"github.com/crate-crypto/go-ipa/bandersnatch/fr"
+	"github.com/ethereum/go-ethereum/common/lru"
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/gballet/go-verkle"
+	"github.com/holiman/uint256"
+)
+
+const (
+	// The spec of verkle key encoding can be found here.
+	// https://notes.ethereum.org/@vbuterin/verkle_tree_eip#Tree-embedding
+	VersionLeafKey    = 0
+	BalanceLeafKey    = 1
+	NonceLeafKey      = 2
+	CodeKeccakLeafKey = 3
+	CodeSizeLeafKey   = 4
+)
+
+var (
+	zero                                = uint256.NewInt(0)
+	verkleNodeWidthLog2                 = 8
+	headerStorageOffset                 = uint256.NewInt(64)
+	mainStorageOffsetLshVerkleNodeWidth = new(uint256.Int).Lsh(uint256.NewInt(256), 31-uint(verkleNodeWidthLog2))
+	codeOffset                          = uint256.NewInt(128)
+	verkleNodeWidth                     = uint256.NewInt(256)
+	codeStorageDelta                    = uint256.NewInt(0).Sub(codeOffset, headerStorageOffset)
+
+	index0Point *verkle.Point // pre-computed commitment of polynomial [2+256*64]
+
+	// cacheHitGauge is the metric to track how many cache hit occurred.
+	cacheHitGauge = metrics.NewRegisteredGauge("trie/verkle/cache/hit", nil)
+
+	// cacheMissGauge is the metric to track how many cache miss occurred.
+	cacheMissGauge = metrics.NewRegisteredGauge("trie/verkle/cache/miss", nil)
+)
+
+func init() {
+	// The byte array is the Marshalled output of the point computed as such:
+	//
+	// 	var (
+	//		config = verkle.GetConfig()
+	//		fr     verkle.Fr
+	//	)
+	//	verkle.FromLEBytes(&fr, []byte{2, 64})
+	//	point := config.CommitToPoly([]verkle.Fr{fr}, 1)
+	index0Point = new(verkle.Point)
+	err := index0Point.SetBytes([]byte{34, 25, 109, 242, 193, 5, 144, 224, 76, 52, 189, 92, 197, 126, 9, 145, 27, 152, 199, 130, 165, 3, 210, 27, 193, 131, 142, 28, 110, 26, 16, 191})
+	if err != nil {
+		panic(err)
+	}
+}
+
+// PointCache is the LRU cache for storing evaluated address commitment.
+type PointCache struct {
+	lru  lru.BasicLRU[string, *verkle.Point]
+	lock sync.RWMutex
+}
+
+// NewPointCache returns the cache with specified size.
+func NewPointCache(maxItems int) *PointCache {
+	return &PointCache{
+		lru: lru.NewBasicLRU[string, *verkle.Point](maxItems),
+	}
+}
+
+// get loads the cached commitment, or nil if it's not existent.
+func (c *PointCache) get(addr string) (*verkle.Point, bool) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.lru.Get(addr)
```

Comment:
In general, an `lru` needs full lock, not just `rlock`. A get-operation on an lru is a modifying operation, since something is moved to the front of a queue. 

IMO we can merge this PR anyway, if you first confirm that there's no chance this path will be executed in 'normal' mode